### ISSUE TITLE
Possibility of using Array in place of TypeRepMap

### ISSUE THREAD

**Issue by re-xyr** [re-xyr/cleff#3]
Maybe, we can have a

```Haskell
data Env es = Env
  { envLen :: {-# unpack #-} !Int
  , envArr :: {-# unpack #-} !(SmallArray Any) -- Effect numbers in the stack is usually way less than 128
  }
```

And the index of handler is just length minus the index of the effect in the effect stack.

Slicing, appending, updating and indexing should be supported.

---

Actually, this blocks the previous trick of using `interpret f . raise` for `interpose`, and it must mutate the array now. Oops.

--- 

Note that `runHere` can be directly defined as array mutation then. We can use `vector` to prototype the whole thing first before switching to self-wrapped implementation.

**Comment by re-xyr** [comment]
Maybe this is a better API:

```haskell
insertHdl :: Hdl e -> Env es -> Env (e : es) -- In place mutation
getHdl :: e :> es => Env es -> Hdl e
bracketGrow :: (Env es -> Env (e ++ (f ++ es))) -> Env (g ++ es) -> Env (f ++ es) -- clones only once
```

**Comment by re-xyr** [comment]
Fixed by bea7c56

**Comment by re-xyr** [comment]
Reopening in hope of a better generalized heterogeneous vector API:

```haskell
data HVector (f :: k -> Type) (es :: [k])

empty :: HVector f '[]

cons :: f e -> HVector f es -> HVector f (e : es)
prepend :: KnownList es' -> HVector f es' -> HVector f es -> HVector f (es' ++ es)

uncons :: HVector f (e : es) -> HVector f es
slice :: KnownList es' => HVector f (es' ++ es) -> HVector f es

index :: Elem e es => HVector f es -> f e
take :: Elems es' es => HVector f es -> HVector f es'

modify :: Elem e es => f e -> HVector f es -> HVector f es
batch :: Elems es' es => HVector f es' -> HVector f es -> HVector f es
```
